### ISSUE TITLE
popup still can be dragged off the screen

### ISSUE THREAD

**Issue by balmas** [alpheios-project/components#770]
seen in qa-3.0.0.58/components 1.2.28 

On a quick test in both Chrome and Safari, it seems that this is still a problem. It's better, in that after being dragged offscreen it usually seems to come back after a lookup, but it doesn't bounce back nicely when dragged like the toolbar does.

_Originally posted by @balmas in https://github.com/alpheios-project/components/issues/712#issuecomment-524425998_

**Comment by kirlat** [comment]
Good thing is that I can reproduce this issue in Chrome on some sites (have not tested it in Safari yet). The bad thing is that it's a little complicated.

The way it is done now (i.e. that the popup can be dragged outside of the screen but then bounces back) is to improve performance. The within bounds check is done only once, at the end of the drag movement, and, if the popup is outside of the screen at this moment, it is moved back. If to do this check during each iteration of the popup movement, it will occur hundreds or maybe thousands times more often, producing a significant CPU load.

If to check the popup position in the elements inspector, it is visible that the problem is not with the popup being not moved back (its coordinates are adjusted correctly so that it stays within the viewport), but that it is closed. It has `display: none` at its props after the drag is complete. If set it so that the popup will appear, it will be shown in the correct position.

This occurs due to the clickaway directive attached to the popup. My guess is that a clickaway reacts to the mouse button release event, and since this happens outside of the visible screen area, after the popup position has been adjusted, then it is considered as a click outside of the popup's rectangle that shall close the popup. That's why the toolbar works while the popup not: the toolbar has no clickaway directive attached to it.

What's interesting is that it works in FF without any issues but fails in Chrome.

Removal of a clickaway directive fixes the issue but I think it's not an option.

I will look further into the best way to solve this.

**Comment by kirlat** [comment]
It works in FF and does not work in Chrome because, when a popup is dragged outside of the viewport and released, click target is reported as being the popup's header in FF (and thus it does not close as the click is within the popup area itself). In Chrome, however, the click target is reported as being within the `html` element, which is outside of the popup area and thus the popup closes. I would say FF probably does a better job here.

We cannot filter on the click target alone, unfortunately, because `html` can be a valid click target when popup is within the viewport and not moved.

**Comment by kirlat** [comment]
The best, and probably the simplest solution would be to analyze click coordinates and, if those are outside of the viewport, ignore them. This worked pretty reliably in my tests. The best place to do that would probably be a clickaway directive.

We, however, using some not-so-obvious algorithm to calculate the viewport's bounds. It was working reliably for us so far, but the clickaway directive has no access to that data as of now. To solve this, without duplicating the viewport bounds calculations, we can:
1. Modify a clickaway value to be an object with a callback and viewport dimensions, Then we can decide inside the clickaway directive if the click is within the viewport or not.
2. A click event is returned to a clickaway's callback. Inside the callback's code in the popup module we can use event's `clientX` and `clientY` coordinates to check if the click is within the viewport.

@balmas, @irina060981: What do you think about the approaches above? Which one do you like better? Are there any other solutions to that?

**Comment by irina060981** [comment]
I have made some experiments and as I could see, that if we click inside viewport than the html tree has something more then html - body, divs and so on
But if you drag the popup over the screen and click to drop it, it has only html

![image](https://user-images.githubusercontent.com/12377640/63688992-1fe9da80-c84c-11e9-9c1e-4189f1ffc48f.png)

May be it would be enough to track does click was inside body or (I could expect some extraodinary layout with short body and some divs overflow) inside some other element? 
And if the tree has at the first place html - than it is outside the viewport, because html itself could be some valuable point of click.

@balmas and @kirlat, what do you think?

**Comment by balmas** [comment]
@kirlat and @irina060981 I think either approach is fine. Irina's sounds a little simpler, and thus maybe a little less error prone. let's go with that.

**Comment by kirlat** [comment]
OK, great, will follow with Irina's approach.

**Comment by kirlat** [comment]
I was testing it on several sites in Google Chrome and it seems to me that on some occasions valid clicks within the viewport produces the same path as the ones outside. Here is one of the examples: https://la.wikipedia.org/wiki/Vicipaedia:Pagina_prima. 

If to drag the popup outside of the screen, the path is: [html, document, Window]. But if to click on the lower part of the gray navigational bar at the left (see the position of a cursor on a screenshot below), the path is the same:
![image](https://user-images.githubusercontent.com/18631055/63763940-c975ae00-c8d6-11e9-8b88-6f4414573b6e.png)
So I'm not sure if we can use path only to identify out-of-the-viewport clicks. Will do more tests.

**Comment by irina060981** [comment]
I have tested on this Wikipedia page - and I see the correct path on the same place inside Chrome

![image](https://user-images.githubusercontent.com/12377640/63764977-67836b80-c90b-11e9-9054-d41c6656c1eb.png)

How I could reproduce the same html tree? 
@kirlat, what exact steps did you do?

**Comment by kirlat** [comment]
Hmm, that's strange. Here is what I do: load the page, activate the plugin, and double click on the word (I was clicking on `summum` from the Pagina mensualis article but it probably does not matter). The popup will appear. Then, without clicking anywhere else, move the mouse cursor to the aforementioned gray area (some way down the nav items) and do a single click on the gray area. The popup will disappear.

Here is what I've got in logs for that (I print both `ev.target` and `path` in a `clickaway.js` after line 70):
![image](https://user-images.githubusercontent.com/18631055/63766066-72260c80-c8db-11e9-9f90-93581ef853f3.png)
@irina060981, how does this work for you? I see in your logs that you have `body` in the path list and I don't. What I print is the path calculated as `const path = ev.path || (ev.composedPath ? ev.composedPath() : undefined)` that clickaway uses. I'm wondering if we're printing slightly different things and if that makes a difference.

P.S. My version of Chrome is 76.0.3809.132.

**Comment by irina060981** [comment]
I have it this way
```
const path = ev.path || (ev.composedPath ? ev.composedPath() : undefined)    
console.info('**************click path', path)
```

If I simply click without lookup (under the last text)
![image](https://user-images.githubusercontent.com/12377640/63766694-1a08fd80-c90f-11e9-8e88-680da8e04731.png)

If I make lookup , move there a popup and drop
![image](https://user-images.githubusercontent.com/12377640/63766772-3efd7080-c90f-11e9-8151-10f37bbc55c6.png)

And it bounces up

**Comment by kirlat** [comment]
Hmm... So, if you do the following sequence (described in the post above): 
> load the page, activate the webextension, double click on the word (I was clicking on summum from the Pagina mensualis article). The popup will appear. Then, without clicking anywhere else, move the mouse cursor to the aforementioned gray area (some way down the nav items) and do a single click on the gray area. The popup will disappear.
will you get [body, html, document, Window]? What version of Chrome and on what platform are you testing it with?

It seems that we're printing the same value (`path` from a clickaway) yet we're getting different results for the same test. Very strange.

**Comment by kirlat** [comment]
@balmas, @irina06098: I've made a branch with the diagnostic messages I've used for my testing `comp-i770-popup-drag`. Can you please, when you will have time, run it inside a webextension and make the test described in https://github.com/alpheios-project/components/issues/770#issuecomment-525241840 and see what the path will be? I would really love to understand the reason for this, it can be important to get the consistent results across environments. Thanks!

**Comment by irina060981** [comment]
Let me try too - may be there are some difference
because I added only console.info

**Comment by irina060981** [comment]
I have a correct  html tree even with Kirill's branch

![image](https://user-images.githubusercontent.com/12377640/63772845-4b88c580-c91d-11e9-8281-fc720cfd669c.png)


It is really strange

**Comment by irina060981** [comment]
It is outside the popup
![image](https://user-images.githubusercontent.com/12377640/63772938-7e32be00-c91d-11e9-91a8-e6f20ef07541.png)

**Comment by kirlat** [comment]
@irina060981, what version of Chrome do you have? Do you run it on Windows 10?

**Comment by irina060981** [comment]
Version 76.0.3809.100 (Official Build) (64-bit)
Windows 10

**Comment by irina060981** [comment]
I think it is almost the same as yours

**Comment by kirlat** [comment]
> I think it is almost the same as yours
Truly weird. I'm wondering if there are any plugins that modify a page layout. Those ones can make the path differ.

**Comment by kirlat** [comment]
> should we use coordinates instead?

I will do this change

**Comment by monzug** [comment]
texts-test.alpheios.net  only in safari in full screen

**Comment by kirlat** [comment]
In texts-test.alpheios.net we're using an embed-lib version of the Alpheios. Is it the same version we were doing the webextension tests with?

**Comment by balmas** [comment]
hmm. I edited this item but my comment seems to have been lost. trying again.

I have been able to reproduce the problem in Safari with the alpheios texts environment, only on the pages with text not on the collections navigation pages.  I think it might be because of the way we scroll the text -- as you scroll down, the text increases.

Most of the time the popup comes back into the viewport with the next lookup, but I have also seen that it doesn't and it requires a refresh of the browser to get it back.  @monzug I would like to know if you think we could live with this and put it in the FAQ. It seems to me to be a bit of an edge case -- users hopefully normally won't be trying to drag the popup off the screen. Let me know if you think it can be deferred. Thanks!
